misc/python/materialize/cli/scratch.py

```diff
@@ -0,0 +1,304 @@
+
+from materialize import git
+from materialize import ui
+from materialize import spawn
+from materialize import ssh
+
+
+def check_required_vars() -> None:
+    """Set reasonable default values for the
+    environment variables necessary to interact with AWS."""
+    if not os.environ.get("AWS_PROFILE"):
+        os.environ["AWS_PROFILE"] = "mz-scratch-admin"
+    if not os.environ.get("AWS_DEFAULT_REGION"):
+        os.environ["AWS_DEFAULT_REGION"] = "us-east-2"
+
+
+SPEAKER = ui.speaker("scratch> ")
+ROOT = os.environ["MZ_ROOT"]
+
+EC2: Any  # "boto3.resources.factory.ec2.ServiceResource"
+
+
+def launch(
+    *,
+    key_name: str,
+    instance_type: str,
+    ami: str,
+    tags: Dict[str, str],
+    display_name: Optional[str] = None,
+    subnet_id: Optional[str] = None,
+    size_gb: int,
+    security_group_id: str,
+    nonce: str,
+) -> Any:
+    """Launch and configure an ec2 instance with the given properties."""
+
+    if display_name:
+        tags["Name"] = display_name
+    tags["scratch-created"] = str(datetime.now().timestamp())
+    tags["nonce"] = nonce
+    tags["git_ref"] = git.describe()
+
+    SPEAKER(f"launching instance {display_name or '(unnamed)'}")
+    with open(ROOT + "/misc/load-tests/provision.bash") as f:
+        provisioning_script = f.read()
+    i = EC2.create_instances(
+        MinCount=1,
+        MaxCount=1,
+        ImageId=ami,
+        InstanceType=instance_type,
+        KeyName=key_name,
+        UserData=provisioning_script,
+        TagSpecifications=[
+            {
+                "ResourceType": "instance",
+                "Tags": [{"Key": k, "Value": v} for (k, v) in tags.items()],
+            }
+        ],
+        NetworkInterfaces=[
+            {
+                "AssociatePublicIpAddress": True,
+                "DeviceIndex": 0,
+                "Groups": [security_group_id],
+                "SubnetId": subnet_id,
+            }
+        ],
+        BlockDeviceMappings=[
+            {
+                "DeviceName": "/dev/sda1",
+                "Ebs": {
+                    "VolumeSize": size_gb,
+                    "VolumeType": "gp3",
+                },
+            }
+        ],
+    )[0]
+
+    return i
+
+
+async def setup(i: Any, subnet_id: Optional[str]) -> None:
+    def is_ready(i: Any) -> bool:
+        return bool(
+            i.public_ip_address and i.state and i.state.get("Name") == "running"
+        )
+
+    done = False
+    async for remaining in ui.async_timeout_loop(60, 5):
+        SPEAKER(f"Waiting for instance to become ready: {remaining}s remaining")
+        i.reload()
+        if is_ready(i):
+            done = True
+            break
+
+    if not done:
+        raise RuntimeError(
+            f"Instance {i} did not become ready in a reasonable amount of time"
+        )
+
+    done = False
+    async for remaining in ui.async_timeout_loop(180, 5):
+        try:
+            ssh.runv(["[", "-f", "/DONE", "]"], "ubuntu", i.public_ip_address)
```

Comment:
I'm a little confused about this. Isn't this blocking? How does that play with Python's asyncio? TBH might be simpler to use threads but I don't have much fo this paged in.

Reply:
Yes it's blocking but that's not a big deal in practice since it typically finishes quickly. Yielding in between tries is good enough here.

I could also use threads -- I don't have a strong opinion.